### ISSUE TITLE
Verifying linked data graphs

### ISSUE THREAD

**Issue by onthebreeze** [uncefact/spec-untp#10]
We've an important UNTP section - https://uncefact.github.io/spec-untp/docs/specification/TrustGraphs that needs quite a bit of discussion about how we should even think about doing this. 

basically the problem is that links between related things can be technically valid but business invalid.  Some examples.

* A credential issuer claims to be ABN 3413288567 but the linked  government identity credential subject is ABN 3455667788.  So the issuer is lying about their business identity.
* A conformity credential is issued by a certifier and the scope is about motorcycle helmet safety.  The conformity credential has a linked accreditation credential from NATA which is technically valid but days the scope of accreditation is animal health.  So the certificate issuer is certifying something they are not authorised to do.
* You scan a barcode with GTIN : 12345678910 and it takes you to a DPP credential about GTIN 10987654321 - so the passport is about the wrong product. 
* and may more like this

all are about verifying a collection of 2 or more credentials and whether the links between then are valid in a **business** sense.

Is it possible to define the validation rules using something like shacl? Or some other way of specifying rules? And if so then who defines the rules? maybe the creator of a UNTP extension for a specific industry / geography like Australian Agriculture?

**Comment by Fak3** [comment]
> A credential issuer claims to be ABN 3413288567 but the linked government identity credential subject is ABN 3455667788. So the issuer is lying about their business identity.

If understand correctly, the validation here must include retrieval of that linked government identity credential. That is already a step that cannot be formalized with just shacl or json-schema rules. 

Some validation steps can be formalized in shacl or json-schema, but full validation process must include other steps for implementers to follow.

**Comment by onthebreeze** [comment]
@Fak3 : yes the mechanism to retrieve a bundle of credentials is separate to the analysis of the graph of linked data that is created from the credentials. Credentials will be discoverable from product or entity identifiers via a link resolution protocol.  And credentials may contain links to other credentials.  But this issue can assume that a verifier has followed links, discovered a number of related credentials, and is holding the data in some kind of graph store  - and now wants to do some verification of the graph.

**Comment by Fak3** [comment]
SHACL playground example validates that issuer of _MotoGearSafetyCredential_ has capability _"caps/MotoGearSafety"_: https://s.zazuko.com/3AJQNCR

**Comment by Fak3** [comment]
One potential issue that comes to mind is that individual subgraphs may contradict each other, and naive validating the result after merging them together conceals where the problem came from.

**Comment by Fak3** [comment]
For example the fraudulent VC can say that its issuer has the needed capability. And this VC is signed by the issuer himself.

If we blindly merge everything that forged VC says with the data (another VC) from the national authority, and then validate, we won't recognize the fraud.

**Comment by Fak3** [comment]
What I'm trying to say is that due to information loss, the majority of checks we have to perform on separate VC subgraphs, and only a few can be done on the merged graph afterwards.

**Comment by onthebreeze** [comment]
The fake accreditation Vc is why we have the trust anchors section of UNTP.  Any VC of type accreditation must be issued by a very short list of known and trusted authorities - eg did:web:NATA.com.Au

**Comment by ashleythedeveloper** [comment]
> Is it possible to define the validation rules using something like shacl? Or some other way of specifying rules?

Based on @Fak3 's example and my research, Yes, it would seem that it is possible to use SHACL to perform the type of validation in the examples you provided with the addition of SPARQL in some cases. But by all means, it's not the only way. 

For example, another option could be using the query language of a graph database like [Cypher Query](https://neo4j.com/developer/cypher/) for neo4j.

```
MATCH (issuer:Issuer)-[:CLAIMS_IDENTITY]->(identity:GovernmentIdentity)
WHERE issuer.abn <> identity.abn
RETURN issuer.abn AS issuerABN, identity.abn AS identityABN`
```

**Comment by nissimsan** [comment]
Related:  https://medium.com/transmute-techtalk/the-united-nations-trust-graph-d65af7b0b678

**Comment by JohnOnGH** [comment]
I think we're going to run into a problem that we all know exists: we cannot expect all items and all nodes in any supply chain to adopt the same standard at the same time (and perhaps, ever). 

This rather clumsy phrase is my way of saying that the existing systems are complex, adaptative, and real-world messy. We cannot expect a single approach to be adopted by all participants, no matter how attractive. Even if we were remarkably optimistic, such adoption would be a gradual rollout, at the beginning no participants use the approach, then some (scattered and in pockets), and then more do until (wildly optimistic), they all do.

That is not to say that I am against this idea or approach. I am a (deep) fan of the graph concept, last year I proposed an approach to consider "Governance" as a "governance graph" concept within Trust over IP's "Governance Architecture Task Force" (https://docs.google.com/presentation/d/1vYUJW76BEK_CQotAZ5maXYwe3H3K3dKCPAcTqEgfGJQ/edit).

My observation is that we have to accept/expect that we will have imperfect information, and we have to decide what to do about that. My expectation is that the relying party / verifier will explore the graph until they have satisfied their need for proof, or until they have exhausted the ability to explore the graph. If exhaustion occurs before acceptable proof, then they need to seek alternative/additional proof and/or accept that the claims are not fully verified and make a decision based on that.

We can consider this in terms of hard rules and soft rules. The hard rules (regulations, law etc.) will demand that we must have proof of claims that are acceptable within the jurisdiction in which they (and we) are being tested. The soft rules might be best efforts, nice to haves, preferences etc. and may allow some "wiggle room".

We need wiggle room.

**Comment by onthebreeze** [comment]
For sure we cannot assume that an entire t-shirt to cotton farm credentials graph exists on day 1, if ever. Our architecture must assume that there are only little snippets of graphs - and that quite often a link to a conformity credential (for example) will take you to a pdf not a vc 

For me I think it's enough to start with just a few minimal use cases where there are just 2 or 3 nodes in a graph. For example :

a product passport links to a conformity credential which links to as accreditation credential. The graph verification should confirm that 
- the product SKU or other identifier in the passport is the same as the one in the conformity credential (ie the certificate is about the right product)
- the attestation scope of the conformity credential is the same as the scope of the linked accreditation (ie the certifier is authorised to issue the atteststion)

I'd suggest that the way forward is to identify several more use cases, create several realistic sample graphs, write a validator (SHACL?) for each - and see if any useful patterns emerge that we can document as best practice / protocols in UNTP

**Comment by onthebreeze** [comment]
For sure we cannot assume that an entire t-shirt to cotton farm credentials graph exists on day 1, if ever. Our architecture must assume that there are only little snippets of graphs - and that quite often a link to a conformity credential (for example) will take you to a pdf not a vc 

For me I think it's enough to start with just a few minimal use cases where there are just 2 or 3 nodes in a graph. For example :

a product passport links to a conformity credential which links to as accreditation credential. The graph verification should confirm that 
- the product SKU or other identifier in the passport is the same as the one in the conformity credential (ie the certificate is about the right product)
- the attestation scope of the conformity credential is the same as the scope of the linked accreditation (ie the certifier is authorised to issue the atteststion)

I'd suggest that the way forward is to identify several more use cases, create several realistic sample graphs, write a validator (SHACL?) for each - and see if any useful patterns emerge that we can document as best practice / protocols in UNTP

**Comment by JohnOnGH** [comment]
Agreed, I was hoping/expecting that we were being pragmatic! The aim is that the linked graph will provide benefit, even if the graph is incomplete.

**Comment by nissimsan** [comment]
Related, here's a link to the demo I presented a couple of meetings back:
https://trace.dpp.ni5.io/

**Comment by nissimsan** [comment]
We need a PR to https://uncefact.github.io/spec-untp/docs/specification/TrustGraphs to close this.

**Comment by nissimsan** [comment]
https://medium.com/transmute-techtalk/the-united-nations-trust-graph-d65af7b0b678

**Comment by nissimsan** [comment]
@zachzeus :
- Same identity across credentials
- Conformity credential linked to claim. Conformity credential is talking about the same thing you are linking
- Certifier is accredited by a trusted 3rd party

**Comment by onthebreeze** [comment]
We can probably use TSM (Towards Sustainable Mining) for this - because Nancy in BC is already working with MAC (Mining Association of Canada) to do exactly that.  So we could make our example 3 patterns real using TSM as a realistic example.

**Comment by zachzeus** [comment]
This really will be explored in our test architecture.  This will be a pull request added once we add the refactored UNTP site.  We are also working on this kind of testing in a reference implementation.  The key outcome for UNTP is that there are some simple well described test for UNTP and implementers will do a lot more.

My next steps are to describe the tests:
Done looks like:

- List of test cases that define actor(s), inputs and outputs and test data (positive and negative).
- Using UNCRM scenario as an example
- Scenario map including prerequisites.

**Comment by philarcher** [comment]
I think SHACL might be a useful component but, for the reasons others have said, it's not a full validation tool. What we can say is that if the graph matches a SHACL pattern then it *might* be OK. It's an early-stage test before you get into what might be  computationally-heavy inspecting of individual claims and the human assessment thereof. 

Should it be helpful in this regard, the [RDF Canonicalization spec](https://www.w3.org/TR/rdf-canon/) is about to become a W3C Rec (WG co-chair's insider knowledge ;-) )

**Comment by zachzeus** [comment]
We've been working on what the trust graph testing for UNTP looks like and this becomes a pattern that implementers can extend.  The UNTP validation will be based on what how the links between the core UNTP schema are validated.  The components that we will demonstrate links for are:
1. DPP
2. DCC
3. DTE
4. DIA

**Comment by onthebreeze** [comment]
@absoludity is playing with RDF processor and https://w3c.github.io/N3/spec/ as a way to process multiple credentials and define cross-credential validation rules. The results will be published as "tier3 testing" minimum viable product.  So assigning this ticket to @absoludity to reference / close when committing his tier3 testing MVP

**Comment by absoludity** [comment]
Good timing :) I've just updated the [POC I've been working on](https://github.com/absoludity/untp-graph-validation-cli) (currently on a personal GH repo, but will move to an org asap), so the README has a good overview of the tooling, architecture and approach that I've been playing with over the past couple of weeks for developing queries of trust graphs, as well as some TODOs which I'll turn into issues once we've repo or otherwise decide on a path forward.

The current code (with the example credentials) demonstrates checking whether all the criteria of the product claims are verified by a DCC:

![Image](https://github.com/user-attachments/assets/f7d6a4ad-6414-4821-af95-c54c662a66bd)

Have a browse of the README there, and if you've time, take it for a spin and try it out. I'll continue with a few of the more obvious items from the TODOs there.

> If understand correctly, the validation here must include retrieval of that linked government identity credential. That is already a step that cannot be formalized with just shacl or json-schema rules.

Though it *can* be done with the notation3 query language that I've settled on in the above POC.

> One potential issue that comes to mind is that individual subgraphs may contradict each other, and naively validating the result after merging them together conceals where the problem came from.

Absolutely - I'd initially tried to use a separate named graph for each credential in the above POC, but hit a few difficulties (but have a TODO to come back to see if it's possible). Named graphs (quads) would allow maintaining provenance rather than just merging all the data, but there may be other ways to handle this too, such as the suggestion above that the user browses and possibly chooses whether to accept an issuer as trustworthy (based on some clear tier 3 test output). One of the next things I want to look at is a query to help decide whether you trust the issuer of a conformance certificate.

**Comment by absoludity** [comment]
Just noting that our [upstream fix to eyereasoner](https://github.com/eyereasoner/eye-js/pull/1586) which stopped manual queries working with the above POC has been merged, and the POC updated to use it now (just re-run `npm install` in the directory).

**Comment by absoludity** [comment]
I've just [merged a PR](https://github.com/absoludity/untp-graph-validation-cli/pull/1) which updates the untp-graph-validation tool so that, rather than running separate queries to validate a requirement (such as whether all criteria of a claim are verified), it instead does two separate steps:
1) First, it runs the (new) configured inference rules in order, **adding** new info to the graph (such as marking claims which have all criteria verified, or trust-dependencies between certificates), before then
2) Running SPARQL queries on the resulting graph+inferences, to pull out the relevant info for each validation for the UX to use.

It's mainly a conception shift to something more inline with what I've understood as the intent (being able to generate inferences on the data, before querying and presenting results etc.). I was also worried about setting up a possible anti-pattern of running separate queries without building up shared understanding (the inferences) and had left myself a TODO to investigate and fix this.

**Comment by absoludity** [comment]
Another merge this week with https://github.com/absoludity/untp-graph-validation-cli/pull/2 so that the tool now verifies the identity trust chains, via a new inference and check. The example is updated with a few DIA's for the DPP and DCC issuers, that all lead back to abr.business.gov.au.

By default, the tier 3 tests will fail because this DID is not trusted by default, but re-running the validation with `--trust-did did:web:abr.business.gov.au` allows the tier 3 test to trust that identity and continue. You can also re-run it after moving one of the other DIAs out of the directory and see how it causes the identity trust chain to be untrusted again.

See the PR for screenshots etc.
